modules/images/regenerate-existing-images/background-process/class-perflab-background-job.php

```diff
@@ -0,0 +1,313 @@
+	 * It also checks if max number of attempts have been attempted for the job.
+	 * For any such condition, it will return false, else true.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool Flag to tell whether this job should run or not.
+	 */
+	public function should_run() {
+		// If job doesn't exist or completed, return false.
+		if ( ! $this->exists() || $this->is_completed() ) {
+			return false;
+		}
+
+		if ( $this->is_running() ) {
+			return false;
+		}
+
+		/**
+		 * Number of attempts to try to run a job.
+		 *
+		 * Repeated attempts may be required to run a failed job. Default 3 attempts.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param int $attempts Number of attempts allowed for a job to run. Default 3.
+		 */
+		$max_attempts = (int) apply_filters( 'perflab_job_max_attempts_allowed', 3 );
+
+		// If number of attempts have been exhausted, return false.
+		if ( $this->get_attempts() >= $max_attempts ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Sets the status of the current job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $status Status to set for current job.
+	 * @return bool Returns true if status is updated, false otherwise.
+	 */
+	public function set_status( $status ) {
+		$valid_statuses = array(
+			self::JOB_STATUS_COMPLETE,
+			self::JOB_STATUS_FAILED,
+			self::JOB_STATUS_QUEUED,
+			self::JOB_STATUS_RUNNING,
+			self::JOB_STATUS_PARTIAL,
+		);
+
+		if ( in_array( $status, $valid_statuses, true ) ) {
+			update_term_meta( $this->id, self::META_KEY_JOB_STATUS, $status );
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Marks the job as completed.
+	 *
+	 * 1. Mark the job status as completed.
+	 * 2. It will also save the timestamp (in seconds) at which the job was completed.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function complete() {
+		$this->set_status( self::JOB_STATUS_COMPLETE );
+		// If job is complete, set the timestamp at which it was completed.
+		update_term_meta( $this->id, self::META_KEY_JOB_COMPLETED_AT, time() );
+	}
+
+	/**
+	 * Marks the job as queued.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function queued() {
+		$this->set_status( self::JOB_STATUS_QUEUED );
+	}
+
+	/**
+	 * Retrieves the job status from its meta information.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Job status.
+	 */
+	public function get_status() {
+		$status = (string) get_term_meta( $this->id, self::META_KEY_JOB_STATUS, true );
+
+		if ( empty( $status ) ) {
+			$status = self::JOB_STATUS_QUEUED;
+		}
+
+		return $status;
+	}
+
+	/**
+	 * Returns the number of attempts executed for a job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return int Number of times the job has been attempted.
+	 */
+	public function get_attempts() {
+		$attempts = get_term_meta( $this->id, self::META_KEY_JOB_ATTEMPTS, true );
+
+		return (int) $attempts;
+	}
+
+	/**
+	 * Set the start time of job.
+	 * It tells at what point of time the job has been started.
```

Comment:
Add line break between Summary and Description.

```suggestion
	 * Set the start time of job.
	 *
	 * It tells at what point of time the job has been started.
```